src/core/crypto/keyless.ts

```diff
@@ -0,0 +1,453 @@
+    const jwtHeader = deserializer.deserializeStr();
+    const expiryDateSecs = deserializer.deserializeU64();
+    const ephemeralPublicKey = EphemeralPublicKey.deserialize(deserializer);
+    const ephemeralSignature = EphemeralSignature.deserialize(deserializer);
+    return new KeylessSignature({
+      jwtHeader,
+      expiryDateSecs,
+      ephemeralCertificate,
+      ephemeralPublicKey,
+      ephemeralSignature,
+    });
+  }
+
+  static isSignature(signature: Signature): signature is KeylessSignature {
+    return signature instanceof KeylessSignature;
+  }
+}
+
+/**
+ * A container for a signature that is a ZeroKnowledgeSig.  Can be expanded to support OpenIdSignature.
+ */
+export class EphemeralCertificate extends Signature {
+  public readonly signature: Signature;
+
+  /**
+   * Index of the underlying enum variant
+   */
+  private readonly variant: EphemeralCertificateVariant;
+
+  constructor(signature: Signature, variant: EphemeralCertificateVariant) {
+    super();
+    this.signature = signature;
+    this.variant = variant;
+  }
+
+  /**
+   * Get the public key in bytes (Uint8Array).
+   *
+   * @returns Uint8Array representation of the public key
+   */
+  toUint8Array(): Uint8Array {
+    return this.signature.toUint8Array();
+  }
+
+  serialize(serializer: Serializer): void {
+    serializer.serializeU32AsUleb128(this.variant);
+    this.signature.serialize(serializer);
+  }
+
+  static deserialize(deserializer: Deserializer): EphemeralCertificate {
+    const variant = deserializer.deserializeUleb128AsU32();
+    switch (variant) {
+      case EphemeralCertificateVariant.ZkProof:
+        return new EphemeralCertificate(ZeroKnowledgeSig.deserialize(deserializer), variant);
+      default:
+        throw new Error(`Unknown variant index for EphemeralCertificate: ${variant}`);
+    }
+  }
+}
+
+/**
+ * A representation of a Groth16 proof.  The points are the compressed serialization of affine reprentation of the proof.
+ */
+export class Groth16Zkp extends Proof {
+  /**
+   * The bytes of G1 proof point a
+   */
+  a: Uint8Array;
+
+  /**
+   * The bytes of G2 proof point b
+   */
+  b: Uint8Array;
+
+  /**
+   * The bytes of G1 proof point c
+   */
+  c: Uint8Array;
+
+  constructor(args: { a: HexInput; b: HexInput; c: HexInput }) {
+    super();
+    const { a, b, c } = args;
+    this.a = Hex.fromHexInput(a).toUint8Array();
+    this.b = Hex.fromHexInput(b).toUint8Array();
+    this.c = Hex.fromHexInput(c).toUint8Array();
+  }
+
+  serialize(serializer: Serializer): void {
+    serializer.serializeFixedBytes(this.a);
+    serializer.serializeFixedBytes(this.b);
+    serializer.serializeFixedBytes(this.c);
+  }
+
+  static deserialize(deserializer: Deserializer): Groth16Zkp {
+    const a = deserializer.deserializeFixedBytes(32);
+    const b = deserializer.deserializeFixedBytes(64);
+    const c = deserializer.deserializeFixedBytes(32);
```

Comment:
So, one of them is 64 bytes and the others are 32 bytes?  Can we document that up above?  maybe call them g1ProofPoint, g2ProofPoint, g3ProofPoint? Or gp1, gp2, gp3?